File: adapter/harness/blend_engine.py
"""
G-1 — StrategyBlendEngine

Configurable, momentum-bounded strategy weight distribution engine.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable


@dataclass
class BlendRule:
    condition: Callable[[dict], bool]
    adjustments: dict[str, float]
    redistribute_to: list[str] | None
    style_override: str | None = None


def normalize_blend(blend: dict[str, float]) -> dict[str, float]:
    floored = {k: max(0.0, v) for k, v in blend.items()}
    total = sum(floored.values())
    if total == 0:
        return blend
    scale = 100.0 / total
    return {k: v * scale for k, v in floored.items()}


def make_blend_adjuster(
    rules: list[BlendRule],
    blend_key: str = "strategy_blend",
    momentum_cap: float = 10.0,
) -> Callable[[dict], dict]:
    def adjuster(state: dict) -> dict:
        blend = state.get(blend_key)
        if not blend:
            return state

        original = dict(blend)
        current = dict(blend)
        fired_rules: list[BlendRule] = []

        for rule in rules:
            if not rule.condition(state):
                continue

            fired_rules.append(rule)
            freed = 0.0

            for key, delta in rule.adjustments.items():
                if key not in current:
                    continue
                old_val = current[key]
                new_val = max(0.0, old_val + delta)
                freed += old_val - new_val  # actual reduction (0 for positive deltas)
                current[key] = new_val

            if freed > 0 and rule.redistribute_to:
                targets = [k for k in rule.redistribute_to if k in current]
                total_target = sum(current[k] for k in targets)
                if total_target > 0:
                    for k in targets:
                        current[k] += freed * (current[k] / total_target)

        # Momentum cap: clamp total per-key change vs original to ±momentum_cap
        for key in current:
            if key not in original:
                continue
            delta = current[key] - original[key]
            if abs(delta) > momentum_cap:
                current[key] = original[key] + (momentum_cap if delta > 0 else -momentum_cap)
                current[key] = max(0.0, current[key])

        current = normalize_blend(current)

        result = {**state, blend_key: current}

        for rule in fired_rules:
            if rule.style_override is not None:
                result["style_override"] = rule.style_override

        return result

    return adjuster

File: packages/harness/src/primitives/blend-engine.ts
/**
 * G-1 — StrategyBlendEngine
 *
 * Configurable, momentum-bounded strategy weight distribution engine.
 * Port of adapter/harness/blend_engine.py. No domain vocabulary.
 */

export interface BlendRule {
  condition: (state: Record<string, unknown>) => boolean
  adjustments: Record<string, number>
  redistributeTo?: string[]
  styleOverride?: string
}

export function normalizeBlend(blend: Record<string, number>): Record<string, number> {
  const floored: Record<string, number> = {}
  for (const [k, v] of Object.entries(blend)) {
    floored[k] = Math.max(0, v)
  }
  const total = Object.values(floored).reduce((sum, v) => sum + v, 0)
  if (total === 0) return blend
  const scale = 100 / total
  const result: Record<string, number> = {}
  for (const [k, v] of Object.entries(floored)) {
    result[k] = v * scale
  }
  return result
}

export function makeBlendAdjuster(
  rules: BlendRule[],
  blendKey = 'strategy_blend',
  momentumCap = 10,
): (state: Record<string, unknown>) => Record<string, unknown> {
  return function adjuster(state: Record<string, unknown>): Record<string, unknown> {
    const blend = state[blendKey] as Record<string, number> | undefined | null
    if (!blend) return state

    const original: Record<string, number> = { ...blend }
    const current: Record<string, number> = { ...blend }
    const firedRules: BlendRule[] = []

    for (const rule of rules) {
      if (!rule.condition(state)) continue
      firedRules.push(rule)
      let freed = 0

      for (const [key, delta] of Object.entries(rule.adjustments)) {
        if (!(key in current)) continue
        const oldVal = current[key]
        const newVal = Math.max(0, oldVal + delta)
        freed += oldVal - newVal
        current[key] = newVal
      }

      if (freed > 0 && rule.redistributeTo) {
        const targets = rule.redistributeTo.filter((k) => k in current)
        const totalTarget = targets.reduce((s, k) => s + current[k], 0)
        if (totalTarget > 0) {
          for (const k of targets) {
            current[k] += freed * (current[k] / totalTarget)
          }
        }
      }
    }

    // Momentum cap: second pass comparing each key vs original
    for (const key of Object.keys(current)) {
      if (!(key in original)) continue
      const delta = current[key] - original[key]
      if (Math.abs(delta) > momentumCap) {
        current[key] = Math.max(0, original[key] + (delta > 0 ? momentumCap : -momentumCap))
      }
    }

    const normalized = normalizeBlend(current)
    const result: Record<string, unknown> = { ...state, [blendKey]: normalized }

    for (const rule of firedRules) {
      if (rule.styleOverride !== undefined) {
        result['style_override'] = rule.styleOverride
      }
    }

    return result
  }
}
